In [1]:
import pandas as pd
import numpy as np
import warnings
from six.moves import urllib


warnings.filterwarnings("ignore")

%matplotlib inline

# Increate the accuracy of the SVC model(Red wine quality)

**Previous accuracy was 59.84%**

In [2]:
data=pd.read_csv("https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv")

In [3]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [4]:
data.quality

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

In [5]:
data.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [8]:
data.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [9]:
X=data.drop("quality",axis=1)
X.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
4      9.4

In [10]:
y=data["quality"]
y.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

In [11]:
from sklearn.model_selection import train_test_split #test data is used to validate data
X_train,X_test,y_train,y_test=train_test_split(
X,y,test_size=0.33,random_state=30)

In [12]:
X_test.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1147           10.0             0.410         0.45             6.2      0.071   
659             7.1             0.840         0.02             4.4      0.096   
871             6.9             0.560         0.03             1.5      0.086   
1333            9.1             0.775         0.22             2.2      0.079   
1411            6.4             0.470         0.40             2.4      0.071   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1147                  6.0                  14.0  0.99702  3.21       0.49   
659                   5.0                  13.0  0.99700  3.41       0.57   
871                  36.0                  46.0  0.99522  3.53       0.57   
1333                 12.0                  48.0  0.99760  3.18       0.51   
1411                  8.0                  19.0  0.99630  3.56       0.73   

      alcohol  
1147     11.8  
659      11.0  
871      10.6  
1333      9.6  
1411     10.6

In [13]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaler.fit(X_train)#evaluate mean and std daviation

StandardScaler()

In [14]:
print(scaler.mean_)

[ 8.3659197   0.52737162  0.27280112  2.49327731  0.08704855 15.21521942
 44.69701214  0.99678087  3.30729225  0.65722689 10.39839714]


In [15]:
X_train_tf=scaler.fit_transform(X_train)
X_train_tf

array([[-0.15399524,  1.41485069, -0.22405969, ...,  0.4153471 ,
        -0.56587839,  0.8580993 ],
       [-0.50145781,  1.35884548, -0.38110676, ..., -0.71065418,
        -0.44947471, -1.14057284],
       [-1.37011425,  0.07072554, -1.11399309, ..., -0.04830049,
        -0.91508943,  0.14428782],
       ...,
       [-0.32772653, -0.04128489, -0.11936164, ...,  0.81275931,
        -0.04206183, -1.33092257],
       [ 0.88839248, -1.44141526,  1.13701493, ..., -0.77688955,
        -0.79868575,  2.00019766],
       [-1.02265168,  1.75088198, -1.27104016, ...,  1.40887764,
        -0.56587839,  1.80984793]])

In [16]:
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

In [17]:
X_test_tf=scaler.transform(X_test)
X_test_tf

array([[ 0.94630291, -0.65734225,  0.92761883, ..., -0.64441881,
        -0.97329127,  1.33397362],
       [-0.73309953,  1.75088198, -1.32338918, ...,  0.68028858,
        -0.50767655,  0.57257471],
       [-0.84892039,  0.18273597, -1.27104016, ...,  1.47511301,
        -0.50767655,  0.19187525],
       ...,
       [ 0.42510905, -1.32940483,  0.29943055, ..., -0.31324197,
         1.06377313,  1.23879875],
       [-0.21190567, -1.55342569,  0.82292079, ...,  0.48158247,
        -0.33307103,  1.42914848],
       [ 0.48301948, -1.2173944 ,  0.45647762, ...,  0.15040562,
         1.18017681,  1.52432334]])

In [18]:
from sklearn.svm import SVC
model=SVC()

In [19]:
model.fit(X_train_tf,y_train)

SVC()

In [20]:
model.score(X_train_tf,y_train)

0.6647992530345471

In [21]:
model.score(X_test_tf,y_test)

0.6534090909090909

In [22]:
y_predict=model.predict(X_test_tf)

In [23]:
y_test

1147    7
659     4
871     5
1333    5
1411    6
       ..
894     6
329     5
1002    7
1135    6
1093    7
Name: quality, Length: 528, dtype: int64

In [24]:
from sklearn.metrics import accuracy_score

In [25]:
accuracy_score(y_test,y_predict)

0.6534090909090909

**The maximum accuracy of the model is 65.34%**

# 2. HYPERPARAMETER TUNING(GridSearch CV) 

In [26]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators=10).fit(X_train,y_train)
y_pred=rf.predict(X_test)

In [27]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print(confusion_matrix(y_pred,y_test))
print(accuracy_score(y_pred,y_test))
print(classification_report(y_pred,y_test))

[[  0   0   0   0   0   0]
 [  0   0   2   0   0   0]
 [  3  15 174  52   4   0]
 [  0   6  47 142  25   0]
 [  0   2   2  15  35   4]
 [  0   0   0   0   0   0]]
0.6647727272727273
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         2
           5       0.77      0.70      0.74       248
           6       0.68      0.65      0.66       220
           7       0.55      0.60      0.57        58
           8       0.00      0.00      0.00         0

    accuracy                           0.66       528
   macro avg       0.33      0.33      0.33       528
weighted avg       0.71      0.66      0.68       528



The main parameters used by a Random Forest Classifier are:

criterion = the function used to evaluate the quality of a split.

max_depth = maximum number of levels allowed in each tree.

max_features = maximum number of features considered when splitting a node.

min_samples_leaf = minimum number of samples which can be stored in a tree leaf.

min_samples_split = minimum number of samples necessary in a node to cause node splitting.

n_estimators = number of trees in the ensamble.

In [28]:
rf=RandomForestClassifier(n_estimators=300, criterion='entropy', max_depth=10, min_samples_split=33, min_samples_leaf=5,).fit(X_train,y_train)
y_pred=rf.predict(X_test)
print(confusion_matrix(y_pred,y_test))
print(accuracy_score(y_pred,y_test))
print(classification_report(y_pred,y_test))

[[  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  3  15 171  47   1   0]
 [  0   8  54 155  45   3]
 [  0   0   0   7  18   1]
 [  0   0   0   0   0   0]]
0.6515151515151515
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         0
           5       0.76      0.72      0.74       237
           6       0.74      0.58      0.65       265
           7       0.28      0.69      0.40        26
           8       0.00      0.00      0.00         0

    accuracy                           0.65       528
   macro avg       0.30      0.33      0.30       528
weighted avg       0.73      0.65      0.68       528



Randomized search CV

In [29]:
from sklearn.model_selection import RandomizedSearchCV
n_estimators=[int(x) for x in np.linspace(200,2000,10)]
criterion=['gini','entropy']
max_depth=[int(y) for y in np.linspace(10,50,20) ]
min_samples_split=[int(z) for z in np.linspace(10,100,25)]
min_samples_leaf=[int(p) for p in np.linspace(1,20,5)]                                             
max_features=['auto','log2','sqrt']
grid={
    'criterion':criterion,
    'n_estimators':n_estimators,
    'max_depth':max_depth,
    'min_samples_split':min_samples_split,
    'min_samples_leaf':min_samples_leaf,
    'max_features':max_features
}
print(grid)

{'criterion': ['gini', 'entropy'], 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000], 'max_depth': [10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 31, 33, 35, 37, 39, 41, 43, 45, 47, 50], 'min_samples_split': [10, 13, 17, 21, 25, 28, 32, 36, 40, 43, 47, 51, 55, 58, 62, 66, 70, 73, 77, 81, 85, 88, 92, 96, 100], 'min_samples_leaf': [1, 5, 10, 15, 20], 'max_features': ['auto', 'log2', 'sqrt']}


In [30]:
rs=RandomizedSearchCV(RandomForestClassifier(),grid,cv=5,n_iter=100,verbose=5,n_jobs=-1,random_state=100)
rs.fit(X_train,y_train)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(), n_iter=100,
                   n_jobs=-1,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [10, 12, 14, 16, 18, 20,
                                                      22, 24, 26, 28, 31, 33,
                                                      35, 37, 39, 41, 43, 45,
                                                      47, 50],
                                        'max_features': ['auto', 'log2',
                                                         'sqrt'],
                                        'min_samples_leaf': [1, 5, 10, 15, 20],
                                        'min_samples_split': [10, 13, 17, 21,
                                                              25, 28, 32, 36,
                                                              40, 43, 47, 51,
                                                              55, 58, 62,

In [31]:
rs.best_params_

{'n_estimators': 2000,
 'min_samples_split': 10,
 'min_samples_leaf': 1,
 'max_features': 'auto',
 'max_depth': 16,
 'criterion': 'entropy'}

In [32]:
model=rs.best_estimator_

In [33]:
y_pred=model.predict(X_test)
print(confusion_matrix(y_pred,y_test))
print(accuracy_score(y_pred,y_test))
print(classification_report(y_pred,y_test))

[[  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  2  14 176  39   2   0]
 [  1   9  46 156  31   1]
 [  0   0   3  14  31   3]
 [  0   0   0   0   0   0]]
0.6875
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         0
           5       0.78      0.76      0.77       233
           6       0.75      0.64      0.69       244
           7       0.48      0.61      0.54        51
           8       0.00      0.00      0.00         0

    accuracy                           0.69       528
   macro avg       0.34      0.33      0.33       528
weighted avg       0.74      0.69      0.71       528



GridSearchCV

In [34]:
from sklearn.model_selection import GridSearchCV

In [35]:
params_grid={
    'criterion':[rs.best_params_['criterion']],
    'n_estimators':[rs.best_params_['n_estimators'],rs.best_params_['n_estimators']-200,rs.best_params_['n_estimators']+200,rs.best_params_['n_estimators']+100,rs.best_params_['n_estimators']-100],
    'max_depth':[rs.best_params_['max_depth']],
    'min_samples_split':[rs.best_params_['min_samples_split'],rs.best_params_['min_samples_split']-2,rs.best_params_['min_samples_split']-1,rs.best_params_['min_samples_split']+1,rs.best_params_['min_samples_split']+2,rs.best_params_['min_samples_split']-3,rs.best_params_['min_samples_split']+3],
    'min_samples_leaf':[rs.best_params_['min_samples_leaf'],rs.best_params_['min_samples_leaf']-5,rs.best_params_['min_samples_leaf']+5,rs.best_params_['min_samples_leaf']-3,rs.best_params_['min_samples_leaf']-2,rs.best_params_['min_samples_leaf']-1],
    'max_features':['auto']
    }

In [36]:
gs=GridSearchCV(RandomForestClassifier(),params_grid,cv=3,n_jobs=-1,verbose=5)
gs.fit(X_train,y_train)

Fitting 3 folds for each of 210 candidates, totalling 630 fits


GridSearchCV(cv=3, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'criterion': ['entropy'], 'max_depth': [16],
                         'max_features': ['auto'],
                         'min_samples_leaf': [1, -4, 6, -2, -1, 0],
                         'min_samples_split': [10, 8, 9, 11, 12, 7, 13],
                         'n_estimators': [2000, 1800, 2200, 2100, 1900]},
             verbose=5)

In [37]:
gs.best_params_

{'criterion': 'entropy',
 'max_depth': 16,
 'max_features': 'auto',
 'min_samples_leaf': 1,
 'min_samples_split': 8,
 'n_estimators': 2200}

In [38]:
model1=gs.best_estimator_

In [39]:
y_pred=model1.predict(X_test)
print(confusion_matrix(y_pred,y_test))
print(accuracy_score(y_pred,y_test))
print(classification_report(y_pred,y_test))

[[  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  2  13 177  38   2   0]
 [  1  10  45 157  30   1]
 [  0   0   3  14  32   3]
 [  0   0   0   0   0   0]]
0.6931818181818182
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         0
           5       0.79      0.76      0.77       232
           6       0.75      0.64      0.69       244
           7       0.50      0.62      0.55        52
           8       0.00      0.00      0.00         0

    accuracy                           0.69       528
   macro avg       0.34      0.34      0.34       528
weighted avg       0.74      0.69      0.72       528



# 3. YOU HVAE TO IMPLEMENT SVR(ADDIMISSOIN_PREDICTION)

In [40]:
df=pd.read_csv("https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv")

In [41]:
df.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0           1        337          118                  4  4.5   4.5  9.65   
1           2        324          107                  4  4.0   4.5  8.87   
2           3        316          104                  3  3.0   3.5  8.00   
3           4        322          110                  3  3.5   2.5  8.67   
4           5        314          103                  2  2.0   3.0  8.21   

   Research  Chance of Admit   
0         1              0.92  
1         1              0.76  
2         1              0.72  
3         1              0.80  
4         0              0.65

In [43]:
df.columns

Index(['Serial No.', 'GRE Score', 'TOEFL Score', 'University Rating', 'SOP',
       'LOR ', 'CGPA', 'Research', 'Chance of Admit '],
      dtype='object')

In [44]:
for i in df.columns:
    df.rename(columns={i:i.strip()},inplace=True)
    
print(df.columns)

Index(['Serial No.', 'GRE Score', 'TOEFL Score', 'University Rating', 'SOP',
       'LOR', 'CGPA', 'Research', 'Chance of Admit'],
      dtype='object')


In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Serial No.         500 non-null    int64  
 1   GRE Score          500 non-null    int64  
 2   TOEFL Score        500 non-null    int64  
 3   University Rating  500 non-null    int64  
 4   SOP                500 non-null    float64
 5   LOR                500 non-null    float64
 6   CGPA               500 non-null    float64
 7   Research           500 non-null    int64  
 8   Chance of Admit    500 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 35.3 KB


In [46]:
df.isnull().sum()

Serial No.           0
GRE Score            0
TOEFL Score          0
University Rating    0
SOP                  0
LOR                  0
CGPA                 0
Research             0
Chance of Admit      0
dtype: int64

In [66]:
x=df.drop("Research",axis=1)
x.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0           1        337          118                  4  4.5  4.5  9.65   
1           2        324          107                  4  4.0  4.5  8.87   
2           3        316          104                  3  3.0  3.5  8.00   
3           4        322          110                  3  3.5  2.5  8.67   
4           5        314          103                  2  2.0  3.0  8.21   

   Chance of Admit  
0             0.92  
1             0.76  
2             0.72  
3             0.80  
4             0.65

In [67]:
y=df["Research"]
y.head()

0    1
1    1
2    1
3    1
4    0
Name: Research, dtype: int64

In [453]:
from sklearn.model_selection import train_test_split #test data is used to validate data
x_train,x_test,Y_train,Y_test=train_test_split(
x,y,test_size=0.33,random_state=80)

In [454]:
x_test

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
256         257        309           99                  3  4.0  4.0  8.56   
231         232        319          106                  3  3.5  2.5  8.33   
444         445        324          113                  5  4.0  5.0  9.25   
459         460        329          113                  4  4.0  3.5  9.36   
353         354        300          102                  3  3.5  2.5  8.17   
..          ...        ...          ...                ...  ...  ...   ...   
350         351        318          107                  3  3.0  3.5  8.27   
145         146        320          113                  2  2.0  2.5  8.64   
174         175        321          111                  4  4.0  4.0  8.97   
414         415        321          110                  4  3.5  4.0  8.35   
155         156        312          109                  3  3.0  3.0  8.69   

     Chance of Admit  
256             0.76  
231             0.74  
444             0.92  
459             0.89  
353             0.63  
..               ...  
350             0.74  
145             0.81  
174             0.87  
414             0.72  
155             0.77  

[165 rows x 8 columns]

In [455]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaler.fit(x_train)#evaluate mean and std daviation

StandardScaler()

In [456]:
print(scaler.mean_)

[260.02089552 316.45671642 106.88358209   3.04179104   3.29850746
   3.48358209   8.56874627   0.71850746]


In [457]:
x_train=scaler.fit_transform(x_train)
x_train

array([[-0.51595518,  1.65709575,  1.81559543, ...,  0.01785766,
         1.46302502,  1.53150907],
       [ 1.39682349, -0.57699583, -0.30763714, ...,  0.01785766,
        -0.93825684, -0.06160622],
       [ 0.1646215 , -0.39826851, -0.14431156, ...,  0.01785766,
         0.37153327,  0.66253709],
       ...,
       [ 1.16757661,  0.49536813,  0.83564193, ..., -1.06983621,
         0.75775343,  0.08322244],
       [-0.50879121,  1.2996411 ,  1.32561868, ...,  0.01785766,
         1.32868758,  1.53150907],
       [-0.60192276,  0.3166408 ,  0.67231635, ...,  0.01785766,
         0.50587071,  0.95219442]])

In [458]:
from sklearn.svm import SVC
model3=SVC()

In [459]:
model3.fit(x_train,Y_train)

SVC()

In [460]:
model3.score(x_train,Y_train)

0.7791044776119403

In [461]:
x_test=scaler.transform(x_test)

In [462]:
x_test

array([[-0.02164159, -0.6663595 , -1.28759064, ...,  0.5617046 ,
        -0.01468689,  0.30046544],
       [-0.20074072,  0.22727714, -0.14431156, ..., -1.06983621,
        -0.40090705,  0.15563677],
       [ 1.32518384,  0.67409545,  0.99896752, ...,  1.64939846,
         1.14397359,  1.45909474],
       ...,
       [-0.60908673,  0.40600446,  0.67231635, ...,  0.5617046 ,
         0.67379252,  1.09702308],
       [ 1.11026489,  0.40600446,  0.50899077, ...,  0.5617046 ,
        -0.36732269,  0.01080811],
       [-0.74520206, -0.39826851,  0.34566519, ..., -0.52598927,
         0.20361146,  0.37287977]])

In [463]:
y_predict=model3.predict(x_test)

In [464]:
Y_test

256    0
231    1
444    1
459    1
353    0
      ..
350    1
145    1
174    1
414    1
155    0
Name: Research, Length: 165, dtype: int64

In [465]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test,y_predict)

0.8121212121212121

The maximum accuracy of the model is 81.21%